File: baekjoon/graph/1707/bipartite_graph.cpp
#include <iostream>
#include <queue>
#include <vector>

typedef std::vector<int> t_vec;
typedef std::queue<int> t_queue;

enum e_color
{
    EMPTY = 0,
    BLACK = 1,
    WHITE = 2
};

int nextIs[] = {BLACK, WHITE};

int *colors;
t_vec *links;

bool bfs(int start)
{
    t_queue q;

    q.push(start);
    colors[start] = BLACK;
    while (!q.empty())
    {
        int &cur = q.front();
        int &curColor = colors[cur];

        for (int &next: links[cur])
        {
            int &nextColor = colors[next];
            if (nextColor == EMPTY)
            {
                nextColor = nextIs[curColor == BLACK];
                q.push(next);
            }
            else if (nextColor == curColor)
                return false;
        }
        q.pop();
    }
    return true;
}

int main(void)
{
    std::cin.tie(0)->sync_with_stdio(0);

    int test_cnt;
    int vCnt, eCnt;
    int from, to;
    const std::string msg[] = {"NO", "YES"};

    std::cin >> test_cnt;
    while (test_cnt--)
    {
        std::cin >> vCnt >> eCnt;

        colors = new int[vCnt + 1];
        for (int i = 1; i <= vCnt; ++i)
            colors[i] = EMPTY;

        links = new t_vec[vCnt + 1];
        while (eCnt--)
        {
            std::cin >> from >> to;
            links[from].push_back(to);
            links[to].push_back(from);
        }

        bool isBipartite = true;
        for (int v = 1; v <= vCnt; ++v)
        {
            if (colors[v] == EMPTY && !bfs(v))
            {
                isBipartite = false;
                break ;
            }
        }
        std::cout << msg[isBipartite] << '\n';

        delete[] links;
        delete[] colors;
    }
    return (0);
}

File: baekjoon/graph/1707/bipartite_graph.py
import sys

sys.setrecursionlimit(20001)

def dfs(cur: int, color: int) -> bool:
    global colors

    colors[cur] = color
    for next in links[cur]:
        if colors[next] == EMPTY:
            if not dfs(next, int(not color)):
                return False
        elif color == colors[next]:
            return False
    return True

if __name__ == '__main__':
    readline = lambda: sys.stdin.readline()
    splitIntoInt = lambda: map(int, readline().split())
    EMPTY, BLACK, WHITE = -1, 0, 1

    test_cnt = int(readline())
    while test_cnt:
        v_cnt, e_cnt = splitIntoInt()
        v_cnt += 1
        links = [[] for _ in range(v_cnt)]
        colors = [EMPTY for _ in range(v_cnt)]

        while e_cnt:
            from_, to = splitIntoInt()
            links[from_].append(to)
            links[to].append(from_)
            e_cnt -= 1

        print_msg = "YES"
        for v in range(1, v_cnt):
            if (colors[v] == EMPTY
                and not dfs(v, BLACK)):
                print_msg = "NO"
                break
        print(print_msg)

        test_cnt -= 1
